ql/pricingengines/swaption/blackswaptionengine.hpp

```diff
@@ -227,28 +227,33 @@ namespace QuantLib {
         // account unwanted cashflows. For the moment we add a check avoiding this situation.
         // Furthermore, we take a copy of the underlying swap. This avoids notifying the swaption
         // when we set a pricing engine on the swap below.
-        VanillaSwap swap = *arguments_.swap;
-        const Leg& fixedLeg = swap.fixedLeg();
+        auto swap = arguments_.swap;
+
+        const Leg& fixedLeg = swap->fixedLeg();
         ext::shared_ptr<FixedRateCoupon> firstCoupon =
             ext::dynamic_pointer_cast<FixedRateCoupon>(fixedLeg[0]);
         QL_REQUIRE(firstCoupon->accrualStartDate() >= exerciseDate,
                    "swap start (" << firstCoupon->accrualStartDate() << ") before exercise date ("
                                   << exerciseDate << ") not supported in Black swaption engine");
 
-        Rate strike = swap.fixedRate();
+        Rate strike = swap->fixedRate();
 
         // using the discounting curve
         // swap.iborIndex() might be using a different forwarding curve
-        swap.setPricingEngine(ext::shared_ptr<PricingEngine>(new
-            DiscountingSwapEngine(discountCurve_, false)));
-        Rate atmForward = swap.fairRate();
+        auto engine = ext::make_shared<DiscountingSwapEngine>(discountCurve_, false);
+        ObservableSettings::instance().disableUpdates();
+        swap->setPricingEngine(engine);
+        ObservableSettings::instance().enableUpdates();
```

Comment:
What problem is this solving? Is it #1791?

Reply:
Yes. I can't make a copy of swap which is virtual. Not sure how to prevent the notifications otherwise.